components/compositing/gl.rs

```diff
@@ -2,125 +2,149 @@
+        );
+
+        gl.framebuffer_texture_2d(
+            gl::FRAMEBUFFER,
+            gl::COLOR_ATTACHMENT0,
+            gl::TEXTURE_2D,
+            texture_ids[0],
+            0,
+        );
+
+        gl.bind_texture(gl::TEXTURE_2D, 0);
+
+        let renderbuffer_ids = gl.gen_renderbuffers(1);
+        let depth_rb = renderbuffer_ids[0];
+        gl.bind_renderbuffer(gl::RENDERBUFFER, depth_rb);
+        gl.renderbuffer_storage(
+            gl::RENDERBUFFER,
+            gl::DEPTH_COMPONENT24,
+            width.get() as gl::GLsizei,
+            height.get() as gl::GLsizei,
+        );
+        gl.framebuffer_renderbuffer(
+            gl::FRAMEBUFFER,
+            gl::DEPTH_ATTACHMENT,
+            gl::RENDERBUFFER,
+            depth_rb,
+        );
+
+        Self {
+            gl,
+            framebuffer_ids,
+            renderbuffer_ids,
+            texture_ids,
+        }
+    }
+
+    pub fn framebuffer_id(&self) -> gl::GLuint {
+        *self.framebuffer_ids.first().expect("Guaranteed by new")
     }
-}
 
-pub fn draw_img(
-    gl: &dyn gl::Gl,
-    render_target_info: RenderTargetInfo,
-    x: i32,
-    y: i32,
-    width: FramebufferUintLength,
-    height: FramebufferUintLength,
-) -> RgbImage {
-    let width = width.get() as usize;
-    let height = height.get() as usize;
-    // For some reason, OSMesa fails to render on the 3rd
-    // attempt in headless mode, under some conditions.
-    // I think this can only be some kind of synchronization
-    // bug in OSMesa, but explicitly un-binding any vertex
-    // array here seems to work around that bug.
-    // See https://github.com/servo/servo/issues/18606.
-    gl.bind_vertex_array(0);
-
-    let mut pixels = gl.read_pixels(
-        x,
-        y,
-        width as gl::GLsizei,
-        height as gl::GLsizei,
-        gl::RGB,
-        gl::UNSIGNED_BYTE,
-    );
-
-    gl.bind_framebuffer(gl::FRAMEBUFFER, 0);
-
-    gl.delete_textures(&render_target_info.texture_ids);
-    gl.delete_renderbuffers(&render_target_info.renderbuffer_ids);
-    gl.delete_framebuffers(&render_target_info.framebuffer_ids);
-
-    // flip image vertically (texture is upside down)
-    let orig_pixels = pixels.clone();
-    let stride = width * 3;
-    for y in 0..height {
-        let dst_start = y * stride;
-        let src_start = (height - y - 1) * stride;
-        let src_slice = &orig_pixels[src_start..src_start + stride];
-        (&mut pixels[dst_start..dst_start + stride]).clone_from_slice(&src_slice[..stride]);
+    pub fn bind(&self) {
+        trace!("Binding fbo {}", self.framebuffer_id());
+        self.gl
+            .bind_framebuffer(gl::FRAMEBUFFER, self.framebuffer_id());
     }
 
-    RgbImage::from_raw(width as u32, height as u32, pixels).expect("Flipping image failed!")
+    pub fn unbind(&self) {
+        trace!("Unbinding fbo {}", self.framebuffer_id());
+        self.gl.bind_framebuffer(gl::FRAMEBUFFER, 0);
+    }
+
+    pub fn read(
```

Comment:
I think a name like `read_to_image` describes a bit better what this does.

Reply:
renamed to `read_back_from_gpu` (dbac481ae9c71311621194382125b0a1c37ab820) to make it explicit that this is a readback